File: lecture_37/p6.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& arr,int r,int c)
    {
        int n=arr.size();
        int m=arr[0].size();
        arr[r][c]=0;
        if(c+1<m && arr[r][c+1]==1)
            dfs(arr,r,c+1);
        if(r+1<n && arr[r+1][c]==1)
            dfs(arr,r+1,c);
        if(c-1>=0 && arr[r][c-1]==1)
            dfs(arr,r,c-1);
        if(r-1>=0 && arr[r-1][c]==1)
            dfs(arr,r-1,c);
        
    }
    int numEnclaves(vector<vector<int>>& arr) {
        for(int j=0;j<arr[0].size();j++)
        {
            if(arr[0][j]==1)
                dfs(arr,0,j);
            if(arr[arr.size()-1][j]==1)
                dfs(arr,arr.size()-1,j);
        }
        for(int i=0;i<arr.size();i++)
        {
            if(arr[i][0]==1)
                dfs(arr,i,0);
            if(arr[i][arr[0].size()-1]==1)
                dfs(arr,i,arr[0].size()-1);
        }
        int count=0;
        for(int i=0;i<arr.size();i++)
        {
            for(int j=0;j<arr[0].size();j++)
            {
                if(arr[i][j]==1)
                    count++;
            }
        }
        return count;
    }
};

File: lecture_48/p6.java
/*package whatever //do not write package name here */

import java.util.*;
import java.lang.*;
import java.io.*;

class GFG {
    public static long merge(int[] arr,int l,int m,int r)
    {
        int[] arr1=new int[m-l+1];
        int[] arr2=new int[r-m];
        for(int i=l;i<=m;i++)
        {
            arr1[i-l]=arr[i];
        }
        for(int i=m+1;i<=r;i++)
        {
            arr2[i-(m+1)]=arr[i];
        }
        int i=arr1.length-1;
        int j=arr2.length-1;
        int k=r;
        long count=0;
        while(i>=0 && j>=0)
        {
            if(arr1[i] > arr2[j])
            {
                count+=(j+1);
                arr[k]=arr1[i];
                k--;
                i--;
            }
            else
            {
                arr[k]=arr2[j];
                k--;
                j--;
            }
        }
        while(i>=0)
        {
            arr[k]=arr1[i];
            k--;
            i--;
        }
        while(j>=0)
        {
            arr[k]=arr2[j];
            k--;
            j--;
        }
        return count;
    }
    public static long solve(int[] nums,int l,int h)
    {
        if(l>=h)
            return 0;
        int mid=l+(h-l)/2;
        long count=0;
        count+=solve(nums, l, mid);
        count+=solve(nums, mid+1, h);
        count+=merge(nums,l,mid,h);
        return count;
    }
	public static void main (String[] args) {
		Scanner sc=new Scanner(System.in);
		int t=sc.nextInt();
		while(t-->0)
        {
            int n=sc.nextInt();
            int[] arr=new int[n];
            for(int i=0;i<n;i++)
            {
                arr[i]=sc.nextInt();
            }
            System.out.println(solve(arr,0,n-1));
        }
	}
}